source/extensions/transport_sockets/tls/context_config_impl.cc

```diff
@@ -409,7 +427,9 @@ ServerContextConfigImpl::ServerContextConfigImpl(
 
   if ((config.common_tls_context().tls_certificates().size() +
        config.common_tls_context().tls_certificate_sds_secret_configs().size()) == 0) {
-    throw EnvoyException("No TLS certificates found for server context");
+    if (createHandshaker()->requireCertificates()) {
```

Comment:
It is inefficient to createHandshaker() just for calling `requireCertificates()`, this should be already known at configuration point, so HandshakerFactory should be able to return this, no?

Reply:
Since requireCertificates is a method on the Handshaker class, and not a property configured statically in the custom_listener_handshaker section in the yaml, (I believe) we have to instantiate the class in order to check the property. Do you think we should make this a property of the custom_listener_handshaker config instead, for efficiency?

Reply:
> I don't think you should add it to the config because it's more of a property of the code than of the config. But you could add it to the factory interface instead of the handshaker interface.

+1

Reply:
This was a good idea and let me stop creating excess handshakers all over the place, done.